File: workers/python/worker.py
import os
import redis
import json
import time
import sandbox
import limiter

REDIS_HOST = os.environ.get('REDIS_HOST', 'localhost')
REDIS_PORT = int(os.environ.get('REDIS_PORT', 6379))

r = redis.Redis(host=REDIS_HOST, port=REDIS_PORT, db=0)

QUEUE_KEY = "execution_queue:python"
RESULT_KEY_PREFIX = "result:"

def main():
    print(f"Python Worker started. Connecting to Redis at {REDIS_HOST}:{REDIS_PORT}")

    # Apply resource limits
    limiter.set_limits()

    while True:
        # Blocking pop from queue
        item = r.brpop(QUEUE_KEY, timeout=5)

        if item:
            _, raw_data = item
            invocation_id = None
            try:
                data = json.loads(raw_data)
                invocation_id = data['invocationId']  # Now int64
                code = data['code']
                input_data = data.get('input', {})

                print(f"Processing invocation: {invocation_id}")

                start_time = time.time()
                status, output, logs = sandbox.run_code(code, input_data)
                duration = int((time.time() - start_time) * 1000)

                # Parse output as JSON if possible
                output_json = None
                error_message = ""
                if status == "SUCCESS":
                    try:
                        output_json = json.loads(output) if output else None
                    except json.JSONDecodeError:
                        # If output is not valid JSON, wrap it
                        output_json = {"result": output}
                else:
                    error_message = output

                result = {
                    'invocationId': invocation_id,
                    'status': status,
                    'output': output_json,
                    'outputRaw': output,
                    'errorMessage': error_message,
                    'logs': logs,
                    'durationMs': duration
                }

                # Save result with invocation ID as key
                r.set(RESULT_KEY_PREFIX + str(invocation_id), json.dumps(result), ex=600)
                print(f"Finished invocation: {invocation_id} - {status}")

            except json.JSONDecodeError as e:
                print(f"Error: Invalid JSON in queue: {e}")

            except KeyError as e:
                print(f"Error: Missing required field in request: {e}")
                if invocation_id:
                    error_result = {
                        'invocationId': invocation_id,
                        'status': 'ERROR',
                        'output': None,
                        'outputRaw': '',
                        'errorMessage': f'Invalid request format: missing {e}',
                        'logs': '',
                        'durationMs': 0
                    }
                    r.set(RESULT_KEY_PREFIX + str(invocation_id), json.dumps(error_result), ex=600)

            except Exception as e:
                print(f"Error processing job: {e}")
                if invocation_id:
                    import traceback
                    error_result = {
                        'invocationId': invocation_id,
                        'status': 'ERROR',
                        'output': None,
                        'outputRaw': '',
                        'errorMessage': str(e),
                        'logs': traceback.format_exc(),
                        'durationMs': 0
                    }
                    r.set(RESULT_KEY_PREFIX + str(invocation_id), json.dumps(error_result), ex=600)
                    print(f"Saved error result for invocation: {invocation_id}")

if __name__ == "__main__":
    main()

File: workers/ruby/worker.rb
require 'redis'
require 'json'
require_relative 'sandbox'

REDIS_HOST = ENV['REDIS_HOST'] || 'localhost'
REDIS_PORT = (ENV['REDIS_PORT'] || 6379).to_i

redis = Redis.new(host: REDIS_HOST, port: REDIS_PORT)

QUEUE_KEY = "execution_queue:ruby"
RESULT_KEY_PREFIX = "result:"

puts "Ruby Worker started. Connecting to Redis at #{REDIS_HOST}:#{REDIS_PORT}"

loop do
  begin
    result = redis.brpop(QUEUE_KEY, timeout: 5)
    next unless result

    _, raw_data = result
    invocation_id = nil

    begin
      data = JSON.parse(raw_data)
      invocation_id = data['invocationId']
      code = data['code']
      input_data = data['input'] || {}

      puts "Processing invocation: #{invocation_id}"

      start_time = Time.now
      status, output, logs = Sandbox.run_code(code, input_data)
      duration = ((Time.now - start_time) * 1000).to_i

      output_json = nil
      error_message = ""

      if status == "SUCCESS"
        begin
          output_json = JSON.parse(output) if output && !output.empty?
        rescue JSON::ParserError
          output_json = { "result" => output }
        end
      else
        error_message = output
      end

      result_data = {
        invocationId: invocation_id,
        status: status,
        output: output_json,
        outputRaw: output.to_s,
        errorMessage: error_message,
        logs: logs,
        durationMs: duration
      }

      redis.set("#{RESULT_KEY_PREFIX}#{invocation_id}", result_data.to_json, ex: 600)
      puts "Finished invocation: #{invocation_id} - #{status}"

    rescue JSON::ParserError => e
      puts "Error: Invalid JSON in queue: #{e.message}"

    rescue KeyError => e
      puts "Error: Missing required field: #{e.message}"
      if invocation_id
        error_result = {
          invocationId: invocation_id,
          status: 'ERROR',
          output: nil,
          outputRaw: '',
          errorMessage: "Invalid request format: #{e.message}",
          logs: '',
          durationMs: 0
        }
        redis.set("#{RESULT_KEY_PREFIX}#{invocation_id}", error_result.to_json, ex: 600)
      end

    rescue => e
      puts "Error processing job: #{e.message}"
      if invocation_id
        error_result = {
          invocationId: invocation_id,
          status: 'ERROR',
          output: nil,
          outputRaw: '',
          errorMessage: e.message,
          logs: e.backtrace&.join("\n") || '',
          durationMs: 0
        }
        redis.set("#{RESULT_KEY_PREFIX}#{invocation_id}", error_result.to_json, ex: 600)
        puts "Saved error result for invocation: #{invocation_id}"
      end
    end

  rescue Redis::ConnectionError => e
    puts "Redis connection error: #{e.message}. Retrying in 1 second..."
    sleep 1
  end
end
